tests/weak/tests/uri.rs

```diff
@@ -0,0 +1,29 @@
+use test_weak::Windows::Foundation::Uri;
+use windows::{factory, IActivationFactory, Interface, Result, HRESULT};
+
+// The Uri class supports weak references, so it is used to test a successful downgrade and upgrade.
+#[test]
+fn test_success() -> Result<()> {
+    let strong = Uri::CreateUri("http://kennykerr.ca")?;
+    let weak = strong.downgrade()?;
+
+    assert_eq!(weak.upgrade().unwrap(), strong);
+    drop(strong);
+    assert_eq!(weak.upgrade(), None);
+
+    Ok(())
+}
+
+// The Uri class factory does not support weak references, so it is used to test an unsuccessful downgrade.
+#[test]
+fn test_failure() -> Result<()> {
+    let strong = factory::<Uri, IActivationFactory>()?;
+    let result = strong.downgrade();
+
+    assert_eq!(result.is_err(), true);
+
+    // E_NOINTERFACE is returned because IWeakReferenceSource is not implemented.
+    assert_eq!(result.unwrap_err().code(), HRESULT(0x8000_4002));
```

Comment:
Out of curiosity, is there a reason we still don’t define constants for common HRESULTs?

Reply:
These are defined in the Win32 metadata. We need to be careful not to depend on these as part of code gen to avoid forcing the consuming apps to require the same dependency. In this particular case, its just a local test so we could rely on generated HRESULTs here.